packages/nimble-components/src/label-provider/base/index.ts

```diff
@@ -0,0 +1,57 @@
+import { DesignToken, FoundationElement } from '@microsoft/fast-foundation';
+import { Observable, Subscriber } from '@microsoft/fast-element';
+import { themeProviderTag } from '../../theme-provider';
+
+/**
+ * Base class for label providers
+ */
+export abstract class LabelProviderBase
+    extends FoundationElement
+    implements Subscriber {
+    protected abstract readonly supportedLabels: {
+        [P in keyof LabelProviderBase]?: DesignToken<string>;
+    };
+
+    private themeProvider?: HTMLElement;
+
+    public constructor() {
+        super();
+        Observable.getNotifier(this).subscribe(this);
+    }
+
+    public override connectedCallback(): void {
+        super.connectedCallback();
+        this.initializeThemeProvider();
+    }
+
+    public override disconnectedCallback(): void {
+        this.themeProvider = undefined;
+    }
+
+    public handleChange(
+        _element: LabelProviderBase,
+        property: keyof LabelProviderBase
+    ): void {
+        if (this.supportedLabels[property]) {
+            const token = this.supportedLabels[property]!;
+            let value = this[property] as string | undefined;
+            if (value === undefined) {
+                value = token.getValueFor(this);
+            }
+            token.setValueFor(this, value);
+            if (this.themeProvider) {
+                token.setValueFor(this.themeProvider, value);
+            }
```

Comment:
If the user is deleting the value off the provider, then I think we should just delete the token value off the provider element instead of fetching / calculating and caching it by calculating the value and setting it on the provider. i.e. lean towards the cascade being the source of truth.

```suggestion
            if (value === undefined) {
                value = token.deleteValueFor(this);
            } else {
                token.setValueFor(this, value);
            }
            if (this.themeProvider) {
                token.setValueFor(this.themeProvider, token.getValueFor(this));
            }
```
Edit: Actually after a bit more thought it probably looks different, continue below.

The design token cascade being the source of truth is important because the design token cascade is observable. I think that is missing in the provider base right now. We should be observing the value of the design tokens on the provider.

I think the high level algorithm is:

- Label provider creates observers for all token values it cares about on itself
- If user sets label provider property / attribute, reflect that on the token value for the provider itself (ie set or delete the token value on self but don't reflect to theme provider yet)
- On changes to the token value on provider itself, reflect to theme provider (also do that in connected as needed)

Reply:
I think the test for the above is a cascade of theme providers:

```html
ie start:
<nimble-theme-provider>
    <label-provider-core my-label="test-parent"></label-provider-core>

     <nimble-theme-provider>
         <label-provider-core my-label="test-child"></label-provider-core>
     </nimble-theme-provider>
         <element-in-test>
</nimble-theme-provider>

then clear child:
<nimble-theme-provider>
    <label-provider-core my-label="test-parent"></label-provider-core>

     <nimble-theme-provider>
         <label-provider-core></label-provider-core>
         <element-in-test>
     </nimble-theme-provider>
</nimble-theme-provider>


set parent:
<nimble-theme-provider>
    <label-provider-core my-label="test-parent-new-value"></label-provider-core>

     <nimble-theme-provider>
         <label-provider-core></label-provider-core>
         <element-in-test><!-- Currently element would be stuck with cached value on theme provider instead of observing cascade changes -->
     </nimble-theme-provider>
</nimble-theme-provider>
```

Reply:
Pushed [a commit](https://github.com/ni/nimble/pull/1328/commits/79b8cc9c490a224fbd6646fbbef26d0f676d6548) to try to adopt this proposal.

I ran into multiple issues getting it to work. I think the fact that the label providers are children of the theme provider ends up complicating things quite a bit.

Issues I hit:
- clearing a label property (the `token.deleteValueFor(labelProvider)` code path): FAST has [some code](https://github.com/microsoft/fast/blob/archives/fast-element-1/packages/web-components/fast-foundation/src/design-token/design-token.ts#L650) that recomputes the local token value on delete. If we delete the value from the label provider first, and it's still set on the parent theme provider at this point, it'll compute the wrong value (which would be evident the next time `labelToken.getValueFor` was called).
   - If we add an explicit `token.deleteValueFor(this.themeProvider)` call at the same time / right afterwards, that ends up triggering our change function (from subscribing to the tokens). That would end up setting an explicit value for the label token on the theme provider (which seems wrong)
   - My current fix for this was to clear the token value (`deleteValueFor`) on the label provider, and do the same for (`deleteValueFor`) the theme provider too, and add code to suppress the change handler in that case (kinda ugly)
- what to do in the connectedCallback: I started by saying 'for all supported label tokens, `token.setValueFor(this.themeProvider, token.getValueFor(thisLabelProvider))`'. I don't think that's right because if the label wasn't set locally, that would explicitly set the default label value on the theme provider (so it wouldn't get cascaded values from its parent anymore). I changed it to 'for label properties on this provider that have a set value, ...' and that seem to work. (I added a separate test for this case - nested theme providers and label providers, with no initial values set, and only set a label value on the parent, and verify that it cascades).

If you're around next week you can take a look and see if what I have makes sense / could be simplified, otherwise we can discusss post-quiet-week.

Reply:
Ahhhhhhh okay. So using the token definition on the label provider doesn't work well because it is influencing it's parent and messing with the cascade. I still think the idea that the label provider should behave declaratively makes sense. But what if it's state is just the attribute values instead of the token value.

So same as before:
- Each label-provider has a readonly map of it's propertyName -> label tokens
- The attributes stay simple, they can have a default value of undefined

But now we just care about our attribute state:
- Base class we track for attribute change and just set that on the theme provider if available:
  ```ts
  handleChanged((_element, property) =>{
      if (this.supportedLabels[property]) {
          const token = this.supportedLabels[property];
          const value = this[property];
          if (this.themeProvider) {
              if (value === null || value === undefined) {
                  token.deleteValueFor(this.themeProvider);
             } else {
                  token.setValueFor(this.themeProvider, value);                
            }
          }
      }
  })
  ```
- And in the connectedCallback we reflect the current attribute state
  ```ts
  connectedCallback() {
      super.connectedCallback()
      this.themeProvider = // find it
      if (this.themeProvider) {
        for (const [property, token] of Object.entries(this.supportedLabels)) {
            const value = this[property];
            if (value === null || value === undefined) {
                  token.deleteValueFor(this.themeProvider);
            } else {
                  token.setValueFor(this.themeProvider, value);                
            }
        }
      }
  ```

I think that simplifies things:
- Can just focus on the label provider attribute values and don't need to deal with calculating defaults, just care about what values override the defaults / undefined
- No issue with token values on the cascade. Label Provider doesn't need to be aware of it at all.

I think the only question may be what should happen in disconnected callback of label provider. If the mental model is that the label provider represents just the attribute values then I guess removing the label provider from a theme provider is effectively like all those attributes are removed. So I guess on disconnect it should do the equivalent of those attrs being set to undefined, i.e. deleteValueFor all the attrs on the theme provider in the label provider disconnected callback?

Reply:
Yeah I think that makes sense, and it definitely simplifies the code.

Updated with that approach, which still works with the previous tests too.